Assets/Scripts/Gameplay/GameState/DSLobbyManagementState.cs

```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using BossRoom.Scripts.Shared.Utilities;
+using Unity.Multiplayer.Samples.Utilities;
+using UnityEngine;
+using VContainer;
+
+namespace Unity.Multiplayer.Samples.BossRoom.Server
+{
+    public class DSLobbyManagementState : GameStateBehaviour
+    {
+        [Inject]
+        ConnectionManager m_ConnectionManager;
+        public override GameState ActiveState => GameState.DedicatedServerLobbyManagement;
+
+        protected override void Start()
+        {
+            base.Start();
+
+            // TODO create DGS lobby here, register to matchmaking, etc. This state bypasses the main menu setup users would normally do get in a game
+            // and does its own game setup MTT-4035
+            var address = "0.0.0.0"; // Change this for fancier infrastructure hosting setup where you can listen on different IP addresses. Right now listening on all.
+            var port = 9998;
+
+            // Some quick command line processing.
+            Dictionary<string, string> args = new();
+            foreach (var oneArg in Environment.GetCommandLineArgs())
+            {
+                var keyValue = oneArg.Split('=');
+                args.Add(keyValue[0], keyValue.Length > 1 ? keyValue[1] : null);
+            }
+
+            var portArg = "-port";
+            if (args.ContainsKey(portArg) && !int.TryParse(args[portArg], out port))
+            {
+                DedicatedServerUtilities.Log("failed to parse -port arg: " + args[portArg]);
+            }
+
+            IEnumerator StartServerCoroutine()
+            {
+                DedicatedServerUtilities.Log($"Starting Headless Server, listening on address {address}:{port}");
+                m_ConnectionManager.StartServerIP(address, port); // This will switch to the char select scene once the server started callback has been called
+
+                yield return new WaitForServerStarted(); // Less performant than just the callback, but way more readable than a callback hell.
+
+                // TODO change scene to char select here and do other init. why is it handled by connection manager right now?
+                SceneLoaderWrapper.Instance.AddOnSceneEventCallback();
+                SceneLoaderWrapper.Instance.LoadScene(SceneNames.CharSelect, useNetworkSceneManager: true);
+            }
+            StartCoroutine(StartServerCoroutine());
```

Comment:
Do we ever want to have an ability to clean this up in OnDestroy or something like that?

```
    Coroutine m_MyCoroutineReference;
     
    void Start()
    {
        m_MyCoroutineReference = StartCoroutine(MyRoutine());
     
        // Later to stop it:
        StopCoroutine(m_MyCoroutineReference);
    }
     
    IEnumerator MyRoutine()
    {
        yield return null; // etc.
    }
```

Reply:
I agreed with you, but actually I think it's better if we use a timeout in the ServerWaiter
If StartServer times out, this will throw an exception and we're good
If StartServer doesn't time out, but DSLobbyManagementState gets destroyed for some reason, we want this to fail loudly (since it was supposed to do some scene loading for example, we don't want to eat the exception by stoping the coroutine on destroy).
Will add timeout + test

Reply:
done